src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -448,20 +475,29 @@
         this)
       (a/stop [this]
         (shutdown! app-context)
-        this))))
+        this)
+      (a/restart [this]
+        (try
+          (run-lifecycle-fns app-context s/stop "stop" (reverse ordered-services))
+          (doseq [svc-id (keys services-by-id)] (swap! app-context assoc svc-id {}))
+          (run-lifecycle-fns app-context s/init "init" ordered-services)
+          (run-lifecycle-fns app-context s/start "start" ordered-services)
+          (catch Throwable t
+            (deliver shutdown-reason-promise {:cause :service-error
```

Comment:
Hmmm, one of the behaviors we might see with this as well is that `stop` could end up being called twice on each service - once from within `restart` but then again after the shutdown promise is realized and it calls back into app/stop.

@cprice404 - do you think this will be a big problem?  Not sure how to work around this without making more fundamental changes to have the restart sequence worked into the lifecycle management on the Trapperkeeper main thread or the like...